streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java

```
@@ -55,16 +55,29 @@ public InternalTopicManager(final StreamsKafkaClient streamsKafkaClient, final i
      * If a topic exists already but has different number of partitions we fail and throw exception requesting user to reset the app before restarting again.
      */
     public void makeReady(final Map<InternalTopicConfig, Integer> topics) {
+        int actualReplicationFactor = replicationFactor;
         for (int i = 0; i < MAX_TOPIC_READY_TRY; i++) {
             try {
                 final MetadataResponse metadata = streamsKafkaClient.fetchMetadata();
                 final Map<String, Integer> existingTopicPartitions = fetchExistingPartitionCountByTopic(metadata);
                 final Map<InternalTopicConfig, Integer> topicsToBeCreated = validateTopicPartitions(topics, existingTopicPartitions);
-                streamsKafkaClient.createTopics(topicsToBeCreated, replicationFactor, windowChangeLogAdditionalRetention, metadata);
+                if (metadata.brokers().size() > 0 && metadata.brokers().size() < replicationFactor) {
+                    log.warn("The number of available brokers {} is less than the desired replication " +
+                            "factor for streams internal topics {}. If running in production, consider " +
+                            "increasing the number of available brokers.",
+                        metadata.brokers().size(), replicationFactor);
+                    actualReplicationFactor = metadata.brokers().size();
```

Do we really want to do this? I would strongly prefer to throw an exception to the user!

---

This is consistent with how things like schema registry, proactive support etc handle cases when the number of brokers is less than replication factor.

---

The goal here is to do the right thing when there are enough brokers, not to provide magic when there just aren't enough brokers (e.g., in a test environment).

---

I'd suggest throwing an exception here with the motivation similar to @norwood above. We have seen similar issues with offset topic num.partitions which we do this "min(broker.size, required num.broekrs)" trick and it introduces much more confusions than user-friendly benefits. For unit tests we should just always override these configs.

---

@guozhangwang @mjsax if I understand you right, you want the default 3, with the option for a user to set it to 1, right?
Or you want no change at all to what we currently have (default 1, user can set higher). I don't like the current option since it leads to trouble in production. I'm ok with the first option.

---

I would prefer to keep the current default 1 and mark the parameter importance as "high", indicating that one most likely wants to change the default if going to production. Default values must not be "production ready" settings IMHO (cf. `state.dir`). Default values should give the best "out-of-the-box" experience when getting started with you first "word count" -- ie., local single broker setting.

---

Ok, I cannot keep the default to 1. This is what led to several bugs. It's not great to expect users to set this parameter which streams should be maintaining correctly.

---

@enothereska, as you know, we have changed the behaviour for the offsets topic so that we default to the safe production setting and fail otherwise. That is based on experience, as @guozhangwang said, and seems more relevant than some of the other examples given.

The question then is how to make it easy for development. For the offsets topic, we set the value to 1 in the `server.properties` that is used in the quickstarts, etc. That won't work here. There are a few possible solutions that will be helpful for this and many other configs:

1. Have a config where users can define whether the environment is prod or dev and change the defaults based on that.
2. Provide methods so that a user can get a prod or dev config. For example, `StreamsConfigs.production()` or `StreamConfigs.development()`.
3. Add an enum to the constructor of `StreamConfigs` where users can define if the environment is production or development.

I think I like `3` best. In any case, we don't need to block this PR on the long-term solution. Still, it may be worth figuring out the end state and then a plan on how to get there.

---

This will now require a KIP and will be done in a separate PR.